### DISCUSSION TITLE
How to update Wifi credentials and MQTT topic in a single command

### DISCUSSION THREAD

**Discussion by ramanraja** [arendst/Tasmota#13619] [Q&A] [answered]
I have the need to configure new devices with a custom MQTT topic and wifi credentials through an automated script. I thought it will be easy to set the MQTT topic while I am still connected to the Wifi Manager portal at 192.168.4.1. So I issued the command:
`http://192.168.4.1/cm?cmnd=backlog%20ssid1%20myssid;password1%20mypasswd;fulltopic%20%25prefix%25/USER1234/%25topic%25/`
But this only updates the Wifi credentials, and not the MQTT topic.
If I put the topic first in the backlog, then the topic gets updated, but not the Wifi credentials.
I would expect the complete backlog command to be executed before it reboots. Am I wrong in that assumption?

**Comment by barbudor** [comment]
I would assume what you are doing to be correct.
Have you looked at the console (web or preferably serial)?

**Reply by sfromis** [reply]
Changing any of those settings will cause a restart shortly after. Not sure of details of delays there, but as `Backlog` has a default inter-command delay of 200 ms, try using `Backlog0` instead, to avoid that.

**Comment by meingraham** [comment]
I believe that when performing commands via `Backlog`, all settings are stored before a restart is performed. Alternatively, you may want to try putting the topic commands first and then the Wi-Fi commands.

**Reply by sfromis** [reply]
My experience is that with a long list of commands in Backlog (and long total inter-command delays), it has happened that Tasmota restarted before the whole sequence was executed, like when configuring via Backlog commands. Just confirmed this with a small test. While the number was much larger than 3, the point is that executing all commands before restart is not ensured. Hence my suggestion of using `Backlog0`.

**Reply by ramanraja** [reply]
I just tried backlog0. It still restarts after executing the first command. 
If the topic is the first command, it alone gets executed and wifi settings are ignored. 
If I put ssid1 and password1 at the beginning, it just executed ssid1 and restarted. Password was never updated.

**Reply by sfromis** [reply]
Did a test. Looks like the issue is with using the http interface, which is not seen (and tested) as being "solid". Indeed only the first command is executed. Doing the same via the preferred MQTT method work as expected.

**Comment by lalo-uy** [comment]
but since the MQTT topic is not yet set, how do you send it ?

**Reply by barbudor** [reply]
As soon as mqtt is enabled there should be a topic which default to 'tasmota_%4d' where %4d is a decimal of (MACaddr & 0x1FFF). This topic follows current FullTopic
https://tasmota.github.io/docs/MQTT/#topic

There is also the fallback topic which do not follow Fulltopic
https://tasmota.github.io/docs/MQTT/#fallback-topic

**Comment by ramanraja** [comment]
Yes, but my usecase is different: I have an AWS lambda that serves multiple users through MQTT.  It partitions the topic space with user IDs in the following format:
`cmnd/USER_ID/DEVICE_ID/POWER`
So the topic has to be field-programmable by embedding the user ID into it.
My users are non-technical, so they cannot be expected to find the IP address of the device and connect to its web interface later. 
They are anyway going to set the Wifi SSID and password through my app, so it would be nice if I can automatically take the user ID of the logged in user and send it to Tasmota along the Wifi credentials.
I have to do this in a single operation when the device is still at the known IP 192.168.4.1
After the device gets on the AWS cloud with a default topic like  `cmnd/DEVIE_ID/POWER`, there is no easy way to correlate the device with its owner.

**Comment by lalo-uy** [comment]
So you may need to set all MQTT parameters and etc before the wifi
credentials.
That way the device will restart several times, but keep the 192.16841 IP.

**Comment by Jason2866** [comment]
Tasmota does not support more than one command in Backlog which is triggering a restart.
This command has to be the last one in the Backlog command string.
Maybe the command `WebGetConfig` https://github.com/arendst/Tasmota/pull/13034 is a solution for your use case

**Comment by barbudor** [comment]
Documentation is clearly stating the opposite as 2 exemples out of 3 show multiple commands triggering reboot into the same backlog
https://tasmota.github.io/docs/Commands/#the-power-of-backlog
And the code is clearly intended to delay the restart once after the backlog queue is empty.
This has been working fine with serial and web console since I'm using Tasmota

I'm looking at it right now

@ramanraja Are you using a standard Tasmota binary or a custom build, specially are you using USE_UFILESYS ?

**Reply by Jason2866** [reply]
Ahh, yes. I forgot that this was changed (long time ago). First implementation was the way i wrote

**Reply by ramanraja** [reply]
@barbudor I am using a custom build, with TLS enabled. I am not using USE_UFILESYS.

**Comment by barbudor** [comment, marked as answer]
I'm so stupid. I knew the answer from the beginning but I forgot
There is no bug !

It's just that `;` **must be url_encoded to `%3B`**

`http://192.168.4.1/cm?cmnd=backlog%20ssid1%20myssid%3Bpassword1%20mypasswd%3Bfulltopic%20%25prefix%25/USER1234/%25topic%25/`

Works as expected !

It is even written in the doc 
![image](https://user-images.githubusercontent.com/2996042/141004812-a98e9030-2127-4920-b9bf-5109641a64ac.png)

**Reply by ramanraja** [reply]
@barbudor I was blissfully unaware that my Javascript command `encodeURI(uri) `can let me down so badly :)
I would never have guessed this issue unless you pointed it out. Thanks a ton !
